### DISCUSSION TITLE
Signing JWTs with symmetrical secret key

### DISCUSSION THREAD

**Discussion by PrarthonaPaul** [keycloak/keycloak#24514] [Q&A] [answered]
Hello, 
I am trying to add support for Request and Request URI on an OpenID client that would make authentication requests to Keycloak or other OpenID providers. My understanding is that JWTs can be signed using a symmetric, asymmetric key or not signed at all (plaintext). For asymmetric keys, I was able to upload the public key for a keypair on the Keycloak Console and sign the JWT using that. 

However, for symmetric keys, it seems a bit different. 
First, there is no public key that I can upload and I was not able to just upload the HMAC key on the Client Keys page. 
Next, we need to sign it using a secret key that is shared between the OpenID client and keycloak. I figured that the client secret can be used to do this. However, in Keycloak, the client can be authenticated without using a client secret (Signed JWT or x509 certificate). In that case, there is no client secret I can extract from the console to sign the JWT with. 

![image](https://github.com/keycloak/keycloak/assets/83880425/d9719b92-e684-48f4-be58-251d5ac5eac8)

I noticed under the realm settings in addition to the RSA and the RSA-OAEP public keys, we have 2 key id's specified for AES and  HS256(see the picture below). However, I could not find any way of getting the secret key (programmatically or through the console). 

![Screenshot from 2023-11-02 16-08-49](https://github.com/keycloak/keycloak/assets/83880425/0c323ee3-517a-4dc4-8691-68f6a0292ba0)

The [JWKSUri](http://localhost:8080/realms/master/protocol/openid-connect/certs) for the realm also does not have it. Are there any other secret keys that are shared between the client and Keycloak? Or are we then required to use client-id and client secret authentication when signing the request object using symmetric keys? 
Thanks for all your help.

**Comment by PrarthonaPaul** [comment]
Hi @justin-tay 
Thank you for all your help with answering my other questions. 
Do you think you have any thoughts on this? 
Thanks in advance!

**Comment by justin-tay** [comment, marked as answer]
The `Signed Jwt with Client Secret` authenticator should be used for your scenario.

![HS256](https://github.com/keycloak/keycloak/assets/49700559/163aafee-abaf-4d4f-bb52-2eaf24464b44)

**Reply by PrarthonaPaul** [reply]
Thank you!
